[CARBON-938] Update the DatePicker to use a new library

paulsturgess opened the pull request:
### New Date Picker library
Switched from `react-date-picker` to `react-day-picker` as the old library is no longer being maintained.
- http://react-day-picker.js.org
- https://github.com/gpbl/react-day-picker

![date-picker](https://user-images.githubusercontent.com/4964/29613101-4ded435a-87fc-11e7-8e63-2b6e0ac7ddb1.gif)

### TODO
- [x] Style for disabled days
- [x] Update the nav bar left/right arrows to match design
- [x] Release notes
- [x] Cross browser check of css
- [x] Possibly add a locale switcher to show translated version on Carbon demo
- [x] Fix specs & coverage

### QA
- Check keyboard support is the same (hitting tab closes the picker rather than giving it focus)
- Visit http://localhost:8095/components/date-input?locale=fr to test the picker showing the months in French. Note French is the only locale other than English supported on the Carbon demo site.
- Carbon demo has been updated so maxDate and minDate can be set to test the disabled dates functionality.

SeanArmstrong commented on src/components/date/date.js:
```diff
@@ -1,7 +1,9 @@
 import React from 'react';
 import PropTypes from 'prop-types';
+import moment from 'moment';
```
I tried to decouple moment js from the datepicker incase we ever want to change it for something else.

All date logic should be based in the date helper if possible

paulsturgess replied:
OK thanks for the heads-up I'll refactor

SeanArmstrong commented on src/components/date/date.js:
```diff
@@ -447,27 +450,42 @@ class Date extends React.Component {
+  * @method disabledDays
+  * @return {Array}
+  */
+  disabledDays() {
```
I wouldn't worry about adding too much extra functionality to the datepickers if we don't already support it.

Did we already support this?

paulsturgess replied:
Yeah the maxDate and minDate props were previously added as part of this PR: https://github.com/Sage/carbon/pull/342

paulsturgess commented:
@harpalsingh I've taken the styles you applied in https://github.com/Sage/carbon/pull/1452 and used them to style the new date picker in this PR. Perhaps it's worth you checking this over to make sure you're happy.

SeanArmstrong commented:
![carbon_-_react_components_from_sage](https://user-images.githubusercontent.com/5382335/29656083-1d0a886a-88ab-11e7-9054-5248832078a2.png)

Not sure the alignment is correct. This is on Chrome.

paulsturgess commented:
@SeanArmstrong I've updated this to make the arrows lineup vertically with the nav bar text.

Checked in Chrome, Safari, Firefox, Microsoft IE 11 and Edge.

![screen shot 2017-08-24 at 12 01 56](https://user-images.githubusercontent.com/4964/29663655-124bc2c2-88c4-11e7-9234-2bbbef81ecca.png)

SeanArmstrong commented on src/components/date/navbar/navbar.scss:
```diff
@@ -0,0 +1,50 @@
+  .DayPicker-NavBar {
+    border-radius: 3px 3px 0 0;
+    color: $white;
+    cursor: pointer;
```
This shouldn't be on the whole navbar. Just on the buttons

paulsturgess commented:
As requested by @SeanArmstrong I've tweaked the cursor css on the navbar so now it's just the buttons that use the pointer style, which makes sense as they are the only clickable elements.

![date-picker-hover](https://user-images.githubusercontent.com/4964/29811280-d49e3056-8c9a-11e7-8dd9-1b1fccc56947.gif)

ianrobsonsage commented:
QA Failed.
When testing in application the position of the date pickers is incorrect.
There is a white border around the date picker that should not be there.
When you navigate forward one month todays date is blanked out within the picker.
When you flick between months the blue border around the arrow is not correctly sized.
Have spoken to UI and they have recommended the dates outside of the current month be in a greyed out but selectable format.
Screenshots added to CARBON-953 showing all issues

paulsturgess commented:
The following updates have been made:
- Ensure today's date is no longer white text on white (when outside current month but still visibile)
- Ensure style of selected date is correct (when outside current month but still visible)
- Fixed hover color for outside days (was the wrong blue)
- Changed outside days text color from blue to grey (slightly darker grey than disabled days)
- Fixed focus outline so it is not cropped by the rounded border
- Removed fixed width so that the picker is the correct width if smaller abbreviations are used as headings

![date-picker-hover2](https://user-images.githubusercontent.com/4964/29867336-29e90742-8d73-11e7-82ec-227b3ec57e96.gif)

ianrobsonsage commented:
Re-tested in Chrome, Safari, Firefox, Edge and IE. No issues found. QA Passed

toni-leigh commented:
@ianrobsonsage note this related issue: https://github.com/Sage/carbon/issues/644

paulsturgess commented:
@andrewjtait the arrow style is from the PR @harpalsingh implemented recently here: https://github.com/Sage/carbon/pull/1452

andrewjtait commented:
@paulsturgess ok no worries - just needs the hover state adding then and the css for the `abbr`

paulsturgess commented:
@andrewjtait  A hover state color has been added to the navigation arrows and `abbr` styles have been added to ensure no border or cursor change is applied.

![date-abbr-styles](https://user-images.githubusercontent.com/4964/30032139-123fedb4-918c-11e7-83c4-c9c515920943.gif)
